File: 20.有效的括号/20.php
<?php
class Solution {

    /**
     * @param String $s
     * @return Boolean
     */
    function isValid($s) {
        $arr = [
            ')' => '(',
            ']' => '[',
            '}' => '{'
        ];

        $leng = strlen($s);
        $data = [];
        if ($leng == 0) {
            return true;//判断是否为空
        }else if ($leng % 2 == 1) {
            return false;//判断是否为单数
        }else{
            for ($i = 0; $i< $leng; $i++) {
                if (!isset($arr[$s[$i]])) {
                    // 左括号押入栈
                    $data[] = $s[$i];
                }elseif ($arr[$s[$i]] === end($data)){
                    // 匹配到左括号是栈最顶层到 则弹出
                    array_pop($data);
                }elseif ($arr[$s[$i]] != end($data)) {
                    // 匹配到左括号不是栈最顶层 直接退出循环
                    return false;
                }
            }
        }

        if (empty($data)) {
            return true;//判断数组data是否有值
        }else{
            return false;//
        }
    }
}

File: 20.有效的括号/20.go
package main

func isValid(s string) bool {
	arr := map[byte]byte{'}':'{', ']':'[', ')': '('}

	data := make([]byte, 0)

	length := len(s)
	if length == 0 {
		return true
	} else if length % 2 != 0 {
		return false
	} else {
		for i:= 0; i < length; i++ {
			if _, ok := arr[s[i]]; !ok {
				// 左括号押入栈
				data = append(data, s[i])
			} else if len(data)>0 && arr[s[i]] == data[len(data) - 1] {
				// 匹配到左括号是栈最顶层到 则弹出
				data = data[:len(data) - 1]
			} else {
				// 匹配到左括号不是栈最顶层 直接退出循环
				return false;
			}
		}
	}

	if (len(data) == 0) {
		return true
	} else {
		return false
	}
}